Source/WebKit/UIProcess/wpe/ScreenManagerWPE.cpp

```diff
@@ -0,0 +1,95 @@
+ *    documentation and/or other materials provided with the distribution.
+ *
+ * THIS SOFTWARE IS PROVIDED BY APPLE INC. ``AS IS'' AND ANY
+ * EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+ * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR
+ * PURPOSE ARE DISCLAIMED.  IN NO EVENT SHALL APPLE INC. OR
+ * CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL,
+ * EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO,
+ * PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR
+ * PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY
+ * OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+#include "config.h"
+#include "ScreenManager.h"
+
+#if PLATFORM(WPE) && ENABLE(WPE_PLATFORM)
+#include <WebCore/PlatformScreen.h>
+#include <cmath>
+#include <wpe/wpe-platform.h>
+
+namespace WebKit {
+using namespace WebCore;
+
+PlatformDisplayID ScreenManager::generatePlatformDisplayID(WPEMonitor* monitor)
+{
+    return wpe_monitor_get_id(monitor);
+}
+
+ScreenManager::ScreenManager()
+{
+    // FIXME: default might not be the right display.
+    auto* display = wpe_display_get_default();
+    auto monitorsCount = wpe_display_get_n_monitors(display);
+    for (unsigned i = 0; i < monitorsCount; ++i) {
+        if (auto* monitor = wpe_display_get_monitor(display, i))
+            addMonitor(monitor);
+    }
+
+    g_signal_connect(display, "monitor-added", G_CALLBACK(+[](WPEDisplay*, WPEMonitor* monitor, ScreenManager* manager) {
+        manager->addMonitor(monitor);
+        manager->propertiesDidChange();
+    }), this);
+    g_signal_connect(display, "monitor-removed", G_CALLBACK(+[](WPEDisplay*, WPEMonitor* monitor, ScreenManager* manager) {
+        manager->removeMonitor(monitor);
+        manager->propertiesDidChange();
+    }), this);
+}
+
+ScreenProperties ScreenManager::collectScreenProperties() const
+{
+    WPEMonitor* primaryMonitor = nullptr;
+    ScreenProperties properties;
+    for (const auto& iter : m_monitorToDisplayIDMap) {
+        WPEMonitor* monitor = iter.key;
+        if (!properties.primaryDisplayID && (!primaryMonitor || primaryMonitor == monitor))
+            properties.primaryDisplayID = iter.value;
+
+        auto width = wpe_monitor_get_width(monitor);
+        auto height = wpe_monitor_get_height(monitor);
+
+        ScreenData data;
+        data.screenRect = FloatRect(wpe_monitor_get_x(monitor), wpe_monitor_get_y(monitor), width, height);
+        data.screenAvailableRect = data.screenRect;
+        data.screenDepth = 24;
+        data.screenDepthPerComponent = 8;
+        data.screenSize = { wpe_monitor_get_physical_width(monitor), wpe_monitor_get_physical_height(monitor) };
+        static const double millimetresPerInch = 25.4;
```

Comment:
This can be `constexpr` to make sure it doesn't end in the read-only data section of the library:

```suggestion
        static constexpr double millimetresPerInch = 25.4;
```